File: human_feature_detection/example/sample_3d.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# import rospy
import rclpy
from rclpy.node import Node
from time import sleep
from human_feature_detection_msgs.srv import Feature3d

def main():
    # rospy.init_node("human_feature_detect_sample_3d")
    rclpy.init()
    nd = Node("human_feature_detect_sample_2d_ros")

    # サーバーが立ち上がったらこちら側でクライアントとして定義する
    # service = rospy.ServiceProxy("/human_feature_detection/feature3d", Feature3d)
    client = nd.create_client(Feature3d, "/human_feature_detection/feature3d")

    # 特徴(身長と服の色)を取得してくれるServiceのサーバーが立ち上がるまで待つ
    client.wait_for_service()
    # rospy.wait_for_service("/human_feature_detection/feature3d")

    # ロボット(カメラ)の前方何メートルから何メートルまでの範囲にある点群を検出するかをサーバーに送信
    # 返答結果はresponseに代入される
    # 以下の例では0.2メートルから1.2メートルの範囲の点群で検出する
    req = Feature3d.Request()
    req.min_range = 0.2
    req.max_range = 1.2
    f = client.call_async(req)
    rclpy.spin_until_future_complete(nd, f)
    response = f.result()
    # response = service(0.2, 1.2)

    # 返答されたデータから、テキストで出力する(例)
    print("身長は、およそ" + str(response.height) + "センチメートルです。")
    if (response.color_success):
        print("服の色は、" + str(response.color) + "です。")
    else:
        print("服の色がよくわかりません。")

if __name__ == '__main__':
    main()

File: human_feature_detection_cpp/example/sample_3d.cpp
#include <rclcpp/rclcpp.hpp>
#include <sobits_interfaces/srv/feature3d.hpp>
#include <iostream>

class HumanFeatureDetectClient : public rclcpp::Node {
public:
    HumanFeatureDetectClient() : Node("sample_3d") {
        client_ = this->create_client<sobits_interfaces::srv::Feature3d>("/human_feature_detection_cpp/feature3d");


        // サービスが利用可能になるまで待つ
        while (!client_->wait_for_service(std::chrono::seconds(1))) {
            RCLCPP_INFO(this->get_logger(), "サービスの待機中...");
        }

        // リクエストを作成
        auto request = std::make_shared<sobits_interfaces::srv::Feature3d::Request>();
        request->min_range = 0.2;
        request->max_range = 1.2;

        // 非同期でサービスを呼び出す
        auto future = client_->async_send_request(request);

        // レスポンスを待つ
        if (rclcpp::spin_until_future_complete(this->get_node_base_interface(), future) == rclcpp::FutureReturnCode::SUCCESS) {
            auto response = future.get();
            std::cout << "身長は、およそ " << response->height << " センチメートルです。" << std::endl;
            if (response->color_success) {
                std::cout << "服の色は、" << response->color << " です。" << std::endl;
            } else {
                std::cout << "服の色がよくわかりません。" << std::endl;
            }
        } else {
            RCLCPP_ERROR(this->get_logger(), "サービス呼び出しに失敗しました");
        }
    }

private:
    rclcpp::Client<sobits_interfaces::srv::Feature3d>::SharedPtr client_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<HumanFeatureDetectClient>();
    rclcpp::shutdown();
    return 0;
}
